packages/flutter_tools/lib/src/asset.dart

```diff
@@ -963,13 +1004,73 @@ class ManifestAssetBundle implements AssetBundle {
     result[asset] = variants;
   }
 
+  // Since it is not clear how overlapping asset declarations should work in the
+  // presence of conditions such as `flavor`, we throw an Error.
+  //
+  // To be more specific, it is not clear if conditions should be combined with
+  // or-logic or and-logic, or if it should depend on the specificity of the
+  // declarations (file versus directory). If you would like examples, consider these:
+  //
+  // ```yaml
+  // # Should assets/free.mp3 always be included since "assets/" has no flavor?
+  // assets:
+  //   - assets/
+  //   - path: assets/free.mp3
+  //     flavor: free
+  //
+  // # Should "assets/paid/pip.mp3" be included for both the "paid" and "free" flavors?
+  // # Or, since "assets/paid/pip.mp3" is more specific than "assets/paid/"", should
+  // # it take precedence over the latter (included only in "free" flavor)?
+  // assets:
+  //   - path: assets/paid/
+  //     flavor: paid
+  //   - path: assets/paid/pip.mp3
+  //     flavor: free
+  //   - asset
+  // ```
+  //
+  // Since it is not obvious what logic (if any) would be intuitive and preferable
+  // to the vast majority of users (if any), we play it safe by throwing a `ToolExit`
+  // in any of these situations. We can always loosen up this restriction later
+  // without breaking anyone.
+  void _checkForFlavorConflicts(_Asset newAsset, List<_Asset> previouslyParsedAssets) {
+    bool cameFromDirectoryEntry(_Asset asset) {
+      return asset.originUri.path.endsWith('/');
+    }
+
+    final _Asset? preExistingAsset = previouslyParsedAssets
+        .where((_Asset other) => other.entryUri == newAsset.entryUri)
+        .firstOrNull;
+
+    if (preExistingAsset == null || preExistingAsset.flavor == newAsset.flavor) {
+      return;
+    }
+
+    String errorMessage =
```

Comment:
nit: use a [StringBuffer](https://api.dart.dev/stable/3.2.0/dart-core/StringBuffer-class.html)

Reply:
I checked with #hackers-dart, and I learned that the Dart compiler [does not automatically optimize this in](https://discord.com/channels/608014603317936148/608022273152122881/1174413326231543879). I've replaced `errorMessage` with a `StringBuffer` in [c494656](https://github.com/flutter/flutter/pull/132985/commits/c494656285fcc3e9406f2f61335532c6f46d8395#diff-1339eded5eeb35ab9fbc61051c530c73e307d4640328760c1af0eee36c142be2R1056).

Reply:
https://github.com/dart-lang/sdk/issues/50304#issuecomment-1291705018